Add ERC: Consolidated Dynamic Storage (CDS)

wisecameron opened the pull request:
Reference implementation:
https://github.com/wisecameron/ConsolidatedDynamicStorage/blob/main/contracts/CDSMinimal.sol

wisecameron committed:
Created erc-xxxx: CDS

wisecameron committed:
Update erc-xxxx.md

abcoathup commented on ERCS/erc-xxxx.md:
```diff
@@ -0,0 +1,388 @@
+# ERC-X: Consolidated Dynamic Storage in Solidity  
```
```suggestion
---
eip: 7844
```
Assigning next sequential EIP/ERC/RIP number.
Numbers are assigned by editors & associates.

Please also update the filename.

abcoathup commented on ERCS/erc-xxxx.md:
```diff
@@ -0,0 +1,388 @@
+# ERC-X: Consolidated Dynamic Storage in Solidity  
+
+## Title: Consolidated Dynamic Storage
```
Please fix the formatting for the header.
title:  
description:  
author: 
discussions-to: 
status: Draft
type: Standards Track
category: ERC
created: 
requires:

abcoathup commented on ERCS/erc-xxxx.md:
```diff
@@ -0,0 +1,388 @@
+**Type:** Standards Track  
+**Created:** 2024-13-12  
+**Requires:** None  
+**Discussions-To:** [Ethereum Magicians](https://ethereum-magicians.org/t/erc-xxxx-consolidated-dynamic-storage-cds-new-approach-to-post-deployment-upgradeability/22217/5)
```
```suggestion
discussions-to: https://ethereum-magicians.org/t/erc-7844-consolidated-dynamic-storage-cds/22217
```
Fixed formatting and added assigned number to Eth Magicians title

wisecameron committed:
Update ERCS/erc-xxxx.md

Committing updated header

Co-authored-by: Andrew B Coathup <28278242+abcoathup@users.noreply.github.com>

wisecameron committed:
Updated header, ERC number, file name.

wisecameron committed:
Fixed header section

wisecameron committed:
Fixed label issue.

wisecameron committed:
New label error.

wisecameron committed:
Major overhaul to conform to EIP Walidator

wisecameron committed:
Minor adjustments to content

SamWilsn commented on assets/erc-7844/Adaptability.png:
We have a slight preference for SVGs if you can provide them. Also consider that many of our readers use light mode, so white text on a white background isn't the greatest. You can include a background color, or if you want to be fancy, use CSS media queries inside the SVG.

SamWilsn committed:
Update erc-7844.md

SamWilsn commented on assets/erc-7844/Adaptability.png:
We have a slight preference for SVGs if you can provide them. Also consider that many of our readers use light mode, so white text on a white background isn't the greatest. You can include a background color, or if you want to be fancy, use CSS media queries inside the SVG.

SamWilsn commented on ERCS/erc-7844.md:
```diff
@@ -0,0 +1,391 @@
+
+## Specification 
+
+**Extendable Structs and Storage Spaces**
```
Please use subheadings for these (eg `### Extendable Structs and Storage Spaces`.)

SamWilsn commented on ERCS/erc-7844.md:
````diff
@@ -0,0 +1,391 @@
+}
+```
+
+## Interfaces
````
These should be at level three or deeper (eg. `### Interfaces`)

SamWilsn commented on ERCS/erc-7844.md:
```diff
@@ -0,0 +1,391 @@
+) external;
+
+
+(optional) function put_batch(
```
I'd recommend putting optional functions into separate interfaces for clarity.

SamWilsn commented on ERCS/erc-7844.md:
```diff
@@ -0,0 +1,391 @@
+
+In short, CDS trades initial complexity for **long-term efficiency, safety, and flexibility**—a tradeoff that is more than justified for systems requiring cutting-edge upgradeability. As audited, plug-and-play implementations become available, CDS will become an **accessible and powerful tool** for developers building systems that demand efficiency and capabilities beyond what is currently access ...
+
+**Why Use CDS When Diamond Storage Exists?**
+
+The **Diamond Standard ([ERC-2535](./eip-2535.md))** modularizes smart contract logic elegantly, but retains a rigid storage model. While facets can be upgraded post-deployment, **the storage layout itself remains static**. To mirror the capabilities of CDS, a Diamond storage layer would require its own proxy-delegate solution up-front.  Developers would need to swap the diamond delegate, then re ...
```
This subsection should be in your Motivation section. Rationale should be used to explain choices made within your document, while the Motivation justifies the document as a whole.

SamWilsn commented on ERCS/erc-7844.md:
```diff
@@ -0,0 +1,391 @@
+
+## Reference Implementation
+
+Refer to [CDS Minimal Example](https://github.com/wisecameron/ConsolidatedDynamicStorage/blob/main/contracts/CDSMinimal.sol)
```
Please either include your reference implementation in the `assets/...` directory or remove this section. We do not permit external links except for those listed in EIP-1.

wisecameron replied:
Thank you for taking the time to dig through and find these issues -- will fix each as soon as possible.

wisecameron committed:
Updated ERC -- addressed all review topics.

wisecameron committed:
Merge branch 'wisecameron/erc-cds' of https://github.com/wisecameron/ERCs into wisecameron/erc-cds

wisecameron commented:
Will fix all remaining auto check failures tomorrow

wisecameron committed:
Fix walidator fail

wisecameron committed:
Changing non-standard sections to h3

wisecameron commented:
Changes in progress:
* Put implementation & specification
* Clarifications regarding entryIndex, memberIndex, safeIndex
* Specification uses RFC 2119
* Low-level diagrams depicting storage layout at various points.
* Clarifications regarding what the system is (ie; a central storage and permisision management contract)
* Specification: making basic permission management compulsory.  
* Example use cases
* Optional feature: string alias for memberIndex.
* Clarification regarding intended use(s) of the system.

wisecameron committed:
Major update,
* Updated Contract
* Significantly Improved Hashing Scheme
* New storage-level diagrams
* RFC-2119
* memberIndex, entryIndex, safeIndex, stringIndex clarifications
* Example use cases
* Clear breakdown of the system
* Permission management spec

wisecameron committed:
Small improvements -- updated two images, streamlined a description

wisecameron committed:
Small update -- match storageSpace mask to updated size in dynamic data hash calculation.

wisecameron committed:
Update string storage structure -- store length in its own slot instead of packing with data.

wisecameron commented:
In progress:
- string aliasing 
- put implementation

My plan after this upcoming update is to start promoting the ERC more to encourage greater discussion  before submitting for the next stage.

wisecameron committed:
Added information about lveraging enums to improve readability of CDS syntax

wisecameron committed:
Fully rewrote the following sections:

* Motivation
* Abstract
* Rationale

-- focused on clarifying exact technical benefits, exploring advantages more deeply - added standardization, interoperability

MASDXI committed:
fix: typo and some style

wisecameron committed:
Merge pull request #1 from MASDXI/fix/typo-and-style

fix: some typo and style

SamWilsn commented on ERCS/erc-7844.md:
```diff
@@ -0,0 +1,529 @@
+---
+eip: 7844
+title: Consolidated Dynamic Storage (CDS)
+description: Unbounded in-place storage structure upgradeability managed at the contract level.
```
```suggestion
description: Unbounded in-place storage structure with upgradeability managed at the contract level.
```

Is this more correct? I'm not sure I understand the original sentence.

wisecameron replied:
It's a word salad for sure.  The premise is that it abstracts away manual storage management entirely by allowing people to extend or create mapped storage structures through function invocations (ie; insert_new_member, init_create).  Will revise this.

SamWilsn commented on ERCS/erc-7844.md:
```diff
@@ -0,0 +1,529 @@
+
+## Abstract  
+
+**Consolidated Dynamic Storage (CDS)** introduces a dedicated storage contract that provides dynamically extendable, hash-indexed storage spaces shared across multiple contracts. By consolidating structured storage into a single, logically segmented layer, CDS eliminates slot conflicts, enables deterministic schema expansion, and enforces strict separation between storage and execution.  Consolid ...
```
```suggestion
**Consolidated Dynamic Storage (CDS)** introduces a dedicated storage contract that provides dynamically extendable, hash-indexed storage spaces shared across multiple contracts. By consolidating structured storage into a single, logically segmented layer, CDS eliminates slot conflicts, enables deterministic schema expansion, and enforces strict separation between storage and execution.  Consolidated multi-contract storage management is also offered by [ERC-2535](./eip-2535.md)'s Diamond Pattern.  CDS expands on this concept by introducing consistent cross-contract access and modification procedures, allowing cross-contract dependencies to leverage standardized interfaces that remain stable regardless of logical or data structure updates.  
```

SamWilsn commented on ERCS/erc-7844.md:
```diff
@@ -0,0 +1,529 @@
+`A`, `B`, `C` hash offset structure not only allows CDS to functionally guarantee safety from collisions, but also aid in debugging by 
+allowing maintainers to instantly track the high-level objective of complex storage operations.
+
+**Metadata Root:**  
```
```suggestion
#### Metadata Root
```

Don't be afraid to use more heading levels.

SamWilsn replied:
Re: introductory text, you definitely need some here. You're jumping straight into Yul.

SamWilsn commented on ERCS/erc-7844.md:
```diff
@@ -0,0 +1,529 @@
+| `size`     | 64       | Size of the member in bits.              |
+| `type`     | 64       | Type ID (e.g., `uint256`, `bool`, etc.). |
+
+### Type IDs:
```
```suggestion
### Type IDs
```

SamWilsn commented on ERCS/erc-7844.md:
```diff
@@ -0,0 +1,529 @@
+
+*For arrays, developers **MAY** define unpacking logic to treat `string` or `bytes` fields as indexed collections of dynamic elements.*
+
+### Hash Structure:
```
```suggestion
### Hash Structure
```

SamWilsn commented on ERCS/erc-7844.md:
```diff
@@ -0,0 +1,529 @@
+The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY” and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Extendable Structs and Storage Spaces
+- Extendable structs **MUST** leverage dynamic mappings with deterministic field hashes (`keccak256`).  
```
Don't bold the uppercase keywords. Just leave them uppercase, but otherwise unadorned.

SamWilsn commented on ERCS/erc-7844.md:
```diff
@@ -0,0 +1,529 @@
+![Adaptability](../assets/erc-7844/Adaptability.svg)
+
+The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY” and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
```
I think you could do with a bit more introductory text before jumping straight into requirements. I'm having a bit of a hard time putting the specification into context (though admittedly I'm not a smart contract developer).

wisecameron replied:
Awesome, yeah considering that a lot of the rationale also needs to be moved, I'll do a significant revision and present the narrative in a more straightforward way.  When I initially made this, I was focused too much on the benefits of upgrading storage without re-deployment, when the more fundamental benefit behind all of this is that it's a standardized, holistic approach to storage management in general.  Tertiary benefits like in-place upgradeability, standardized access patterns, etc are all within that fundamental basis, but I don't really communicate this in the proposal.

SamWilsn commented on ERCS/erc-7844.md:
````diff
@@ -0,0 +1,529 @@
+The `init_create` function **MUST** handle storage space creation in both an initialization and live extension setting.  The function **MUST** take an array of types and sizes as input, which **MUST** conform to the above specifications, and be equivalent in length when input.  Developers **MUST** include length equivalency validation for `types` and `sizes`.
+
+```solidity
+function init_create(
````
You should move implementations into the reference implementation section (unless there is _exactly_ one possible implementation). The specification section should define the observable behavior of an implementation, not the internals of how it works.

SamWilsn commented on ERCS/erc-7844.md:
````diff
@@ -0,0 +1,529 @@
+```
+
+## Rationale  
+Proxy-Delegate and Diamond Storage architectures have long been the industry standard for upgradeable smart contracts, providing critical mechanisms for separating logic from state. These approaches have enabled modular contract upgrades, mitigated contract size limitations, and allowed systems to evolve over time. However, as protocols scale, new challenges emerge—managing storage across multipl ...
+
+CDS builds upon these established patterns by introducing a dedicated, structured storage layer that eliminates many of the complexities inherent in Proxy-Delegate and Diamond architectures. In traditional models, storage modifications require careful slot management, pre-planning reserved fields, or governance-heavy migrations to maintain compatibility. These constraints impose operational overh ...
+
+One of the key improvements CDS introduces is standardized storage access across contracts. Whereas Proxy-Delegate and Diamond models require each contract to define its own storage mappings, CDS provides a shared, structured interface that multiple contracts can access consistently. This reduces cross-contract storage fragmentation, simplifies state synchronization, and eliminates the inefficien ...
+
+Furthermore, CDS is designed with long-term system maintainability in mind. As protocols expand, managing state across interdependent contracts becomes increasingly complex. CDS mitigates these risks by externalizing structured storage into a permissioned, extendable layer, ensuring that high-throughput contract interactions remain efficient even as ecosystems grow in scale and complexity.
+
+Proxy-Delegate and Diamond architectures remain valuable for many upgradeable contract designs, offering modularity and logic flexibility. However, for protocols requiring sustainable long-term upgradeability, efficient cross-contract storage, and deterministic schema evolution, CDS provides a more scalable and future-proof alternative. By eliminating migration risks and reducing upgrade complexi ...
````
Almost all of this belongs in the Motivation section, because it's justifying the proposal as a whole. The Rationale section is for explaining choices made within the document itself. My favorite analogy is:

> Motivation: We need to build a shed because...
> Rationale: We chose to paint the shed red because...

wisecameron commented:
@SamWilsn Thank you so much for the thorough review, I sincerely appreciate it and will be sure to implement your recommendations, along with some additional points from the EM discussion.

As an aside, to record where this proposal is, I'm currently building out a lot of external infrastructure (ie; builder's docs, contributor program) around this and HoneyBadger (which is an extended ERC-7844 implementation).  So there will likely be some delay as I'm setting all of this up, but this proposal is still very active.

wisecameron committed:
Update ERCS/erc-7844.md

Co-authored-by: Sam Wilson <57262657+SamWilsn@users.noreply.github.com>

wisecameron committed:
Update ERCS/erc-7844.md

Co-authored-by: Sam Wilson <57262657+SamWilsn@users.noreply.github.com>

wisecameron committed:
Update ERCS/erc-7844.md

Co-authored-by: Sam Wilson <57262657+SamWilsn@users.noreply.github.com>

wisecameron committed:
Abstract statement improvement

wisecameron committed:
Merge branch 'wisecameron/erc-cds' of https://github.com/wisecameron/ERCs into wisecameron/erc-cds

wisecameron committed:
Significant improvements to the motivation section: still a major WIP.

wisecameron committed:
Update -- new motivation content (unfinished), significantly improved architecture diagram)

wisecameron committed:
Improve architecture svg

wisecameron committed:
Architecture image update -- previous had a transparent background

wisecameron committed:
Tightened "Why Standardize" section

wisecameron committed:
Updated "harmoizing upgradeability, control, decentralization"

wisecameron committed:
Updated description

SamWilsn commented on ERCS/erc-7844.md:
```diff
@@ -0,0 +1,548 @@
+created: 2024-12-13
+---
+
+## Abstract  
+
+**Consolidated Dynamic Storage (CDS)** introduces a new foundational storage primitive with broad implications for modern smart contract development.  Its aim is twofold: make storage upgradeability safe and consistent by introducing the concept of *storage abstraction,* and provide a holistic, ready-to-use foundation that naturally fosters good development hygeine, while giving developers space  ...
+
+CDS provides dynamically extendable, hash-indexed storage spaces shared across multiple contracts. By consolidating structured storage into a single, logically segmented layer secured via RBAC and deterministic hash-segregated storage slot derivation, CDS eliminates slot conflicts, enables in-place schema expansion, and enforces strict separation between storage and execution.  It's worth noting  ...
+
+CDS enables in-place storage structure extension, referred to as *storage abstraction,* introducing mapped storage spaces populated by "extendible structs," and thereby removing any need for storage-driven contract redeployments or manual slot management.  Storage spaces can be dynamically allocated in-place, and are logically segmented using an efficient bit-packed hash structure.  By leveraging ...
+
+Additionally, CDS optimizes execution efficiency by enforcing a single-hop storage access model, replacing deep delegation chains with direct, permissioned storage interactions. This reduces redundant SLOAD operations and can lower gas overhead in multi-contract systems. These properties make CDS particularly well-suited for DeFi protocols, DAOs, and modular architectures requiring long-term upgr ...
```
You're gonna need to trim down your abstract a bit. There's a lot of good content here. You certainly give a technical overview of the proposal (which is a requirement), but a good abstract also needs to be concise.

wisecameron replied:
Thank you so much for revisiting the proposal, seriously I really appreciate your time and feedback. Really cool work on EELS by the way, it's been pretty useful for me.

SamWilsn commented on assets/erc-7844/CDSMinimal.sol:
```diff
@@ -0,0 +1,314 @@
+//SPDX-License-Identifier: Unlicensed
```
You'll need to license this under something more permissive. CC0-1.0 is preferred, but MIT/Apache would be okay too.

SamWilsn reviewed (commented):
Few additional comments while you're working on this.

wisecameron committed:
Refine the abstract and permission management sections of the CDS documentation. Enhanced clarity on the capabilities and benefits of Consolidated Dynamic Storage, emphasizing its role as a comprehensive storage abstraction layer. Updated permission management table for improved structure and clarity. Changed SPDX license identifier in the CDSMinimal.sol file to CC0-1.0.

wisecameron commented on ERCS/erc-7844.md:
```diff
@@ -0,0 +1,548 @@
+
+## Abstract  
+
+**Consolidated Dynamic Storage (CDS)** introduces a new foundational storage primitive with broad implications for modern smart contract development.  Its aim is twofold: make storage upgradeability safe and consistent by introducing the concept of *storage abstraction,* and provide a holistic, ready-to-use foundation that naturally fosters good development hygeine, while giving developers space  ...
```
Yeah, definitely better.  Thanks :)

wisecameron committed:
Streamlined abstract, significantly improved motivation section.  Removed flowery language.  Focused on communicating key points with more clarity.

wisecameron committed:
Refine abstract and motivation sections of the CDS documentation for clarity and conciseness. Emphasize the role of CDS as a foundational storage primitive and its benefits in upgradeability and decentralization.

wisecameron committed:
Significantly improved the proposal again -- focused on removing redundancy and enhancing clarity and formatting.

wisecameron committed:
Update ERCS/erc-7844.md

Co-authored-by: Sam Wilson <57262657+SamWilsn@users.noreply.github.com>

wisecameron committed:
Massive update to the rationale section! Thank you so much for the amazing feedback :)

wisecameron committed:
Merge branch 'wisecameron/erc-cds' of https://github.com/wisecameron/ERCs into wisecameron/erc-cds

wisecameron committed:
Added genuine reference implementation, substantially pruned the motivation section, improved interface clarity.

wisecameron committed:
Revised abstract to better highlight core value proposition: in-place creation and extension of mapped structs and storage spaces.

SamWilsn reviewed (commented):
Looking good! Still need to remove the bold/italics from the UPPERCASE keywords, and move example code into a reference implementation section.

wisecameron commented:
> Looking good! Still need to remove the bold/italics from the UPPERCASE keywords, and move example code into a reference implementation section.

Hey thank you man!  

Update:
Been working at a very intense startup since May and haven't had as much time to work on this.  I think it has quite a bit of unique merit but I need to simplify the outward image and build some interest around the concept - planning on building out a website, doing community building, and hosting some hackathons to build interest around it.  Really pleased to see that the proposal hasn't been closed or anything, I 110% intend to get this moving, it's just quite a commitment to take interest from 0 to 1 and I need to be consistently dedicated to pull it off.

SamWilsn commented:
Totally get it! I'm commenting here to bump this off [my queue](https://eipsinsight.com/boards) until you make another change.
